raysect/core/math/function/function3d/tests/test_base.py

```diff
@@ -107,31 +108,79 @@ def test_div_scalar(self):
         with self.assertRaises(ZeroDivisionError, msg="ZeroDivisionError not raised when function returns zero."):
             r(0, 0, 0)
 
-    def test_add_function1d(self):
+    def test_mod_function3d_scalar(self):
+        # Note that Function3D objects work with doubles, so the floating modulo
+        # operator is used rather than the integer one. For accurate testing we
+        # therefore need to use the math.fmod operator rather than % in Python.
+        v = [-1e1, -7, -0.001, 0.00003, 10, 12.3]
+        r1 = 5 % self.f1
+        r2 = self.f1 % -7.8
+        for x in v:
+            for y in v:
+                for z in v:
+                    if self.f1_ref(x, y, z) == 0:
+                        with self.assertRaises(ZeroDivisionError, msg="ZeroDivisionError not raised when function returns 0"):
+                            r1(x, y, z)
+                    else:
+                        self.assertAlmostEqual(r1(x, y, z), math.fmod(5, self.f1_ref(x, y, z)), 15, "Function3D modulo scalar (K % f()) did not match reference function value.")
+                    self.assertAlmostEqual(r2(x, y, z), math.fmod(self.f1_ref(x, y, z), -7.8), 15, "Function3D modulo scalar (f() % K) did not match reference function value.")
+        with self.assertRaises(ZeroDivisionError, msg="ZeroDivisionError not raised when function returns 0"):
+            r1(0, 0, 0)
+        with self.assertRaises(ZeroDivisionError, msg="ZeroDivisionError not raised when modulo scalar is 0"):
+            self.f1 % 0
+
+    def test_pow_function3d_scalar(self):
+        v = [-1e1, -7, -0.001, 0.00003, 10, 12.3]
```

Comment:
looks like the 0 has been dropped from -1e10 compared to other tests. Could you fix this?

Reply:
And similar instances in the other tests.

Reply:
This was deliberate, as pow blows up fairly quickly when large values are used as the exponent. I was getting `OverflowError`s in Python and `nan`s in C when using the original -1e10 value (and the 2.3e49 value at the other end).

I could refactor these tests to use different lists of base and exponent, and keep the exponents small, if you'd prefer.

Reply:
Ah yes of course. I was a little stumped by it was -1e1 rather than -10, so suspected a typo. It can remain as it is. I'll merge it and fix it as I'm just being fussy.